e2e/nomostest/nt.go

```diff
@@ -198,28 +194,124 @@ func RepoSyncNN(ns, name string) types.NamespacedName {
 // DefaultRootRepoNamespacedName is the NamespacedName of the default RootSync object.
 var DefaultRootRepoNamespacedName = RootSyncNN(configsync.RootSyncName)
 
-var sharedNT *NT
+var mySharedNTs *sharedNTs
+
+type sharedNT struct {
+	inUse    bool
+	sharedNT *NT
+	fakeNTB  *testing.FakeNTB
+}
+
+type sharedNTs struct {
+	lock      sync.Mutex
+	testMap   map[string]*sharedNT
+	sharedNTs []*sharedNT
+}
+
+func (snt *sharedNTs) newNT(nt *NT, fake *testing.FakeNTB) {
+	snt.lock.Lock()
+	defer snt.lock.Unlock()
+	snt.sharedNTs = append(snt.sharedNTs, &sharedNT{
+		inUse:    false,
+		sharedNT: nt,
+		fakeNTB:  fake,
+	})
+}
+
+func (snt *sharedNTs) acquire(t testing.NTB) *NT {
+	snt.lock.Lock()
+	defer snt.lock.Unlock()
+	if nt, ok := snt.testMap[t.Name()]; ok {
+		return nt.sharedNT
+	}
+	for _, nt := range snt.sharedNTs {
+		if !nt.inUse {
+			nt.inUse = true
+			snt.testMap[t.Name()] = nt
+			t.Cleanup(func() {
+				snt.release(t.Name())
+			})
+			return nt.sharedNT
+		}
+	}
+	t.Fatal("failed to get shared test environment")
```

Comment:
If all `nt`s are in use, should we wait and retry acquiring an idle shared NT, instead of failing the test?

Reply:
This should never happen unless something else failed or there is a bug, because we create a number of shared test environments equal to the number of test threads. I figured it would be best to fail fast to make it obvious that something went wrong

Reply:
ACK